app/src/main/java/org/openobservatory/ooniprobe/model/database/Url.java

```diff
@@ -56,6 +75,52 @@ public static Url checkExistingUrl(String input, String categoryCode, String cou
 		return url;
 	}
 
+
+	/**
+	 * This function saves or updates input urls ({@code List<Url>}) into the database and returns the
+	 * list of strings ({@code List<String>}) used internally to perform this operation, where each string
+	 * actually useful to continue processing the input URLs.
+	 *
+	 * @param urls
+	 *        List of urls to save.
+	 * @return list of url strings.
+	 */
+	public static List<String> saveOrUpdate(List<Url> urls) {
+
+		Map<String, Url> resultUrlsMap = Maps.uniqueIndex(
+				urls,
+				input -> input.url
+		);
+
+		List<Url> existingUrls = Url.getExistingUrls(resultUrlsMap.keySet());
+
+		List<Url> existingUrlsToUpdate = Lists.newArrayList(Iterables.filter(existingUrls, input -> {
+			Url incomingChanges = resultUrlsMap.get(input.url);
+			if (incomingChanges == null) {
+				return false;
+			}
+			return ((!input.category_code.equals(incomingChanges.category_code)
+					&& !incomingChanges.category_code.equals("MISC"))
+					|| (!input.country_code.equals(incomingChanges.country_code)
+					&& !incomingChanges.country_code.equals("XX")));
+		}));
+
+		if (!existingUrlsToUpdate.isEmpty()) {
+			Url.saveAll(Lists.newArrayList(existingUrlsToUpdate));
+		}
```

Comment:
The code here presents a few issues:
1. At the moment in time in which a certain measurements was run the URL had that category code and country code, so by updating it we are in an unexpected way changing past results.
2. In `resultUrlsMap` you are building a map that is keyed on the URL and using this to lookup if the URL requires updating. This creates an undefined behaviour if the same URL is present in two different country lists, because it would share the same key, yet they are two distinct entries in the test list.

OTOH if we don't update them, since the lookup is done based on the url value, we run the risk of displaying URL entries that are mis-categorised.

I don't really see a way around this, beyond keeping a reference to the URL object throughout the testing lifecycle, so you can use that instead of having to redo a lookup at the end of a test run or better yet just merging the `url` and `measurement` tables together as discussed in: https://github.com/ooni/probe/issues/2024.

Maybe since both of these options have tradeoffs which are more or less equivalent, we should go for the solution that results in the simplest code and hence drop all this logic?

Reply:
I would go for merging the tables. This would obviously need some code refactoring where necessary. 
For this to happen, the `probe-cli` would have to be updated in turn.

If we want to wait for that to happen, we can halt this pull request for that to happen. Otherwise we can ship what we have and when the `probe-cli` is made to support the table merge, we update the apps appropriately

Reply:
Yeah I think we should do it incrementally as the change in probe-cli is going to require quite a bit more work. I suggest we ship this as-is then.